lib/wallet-deposit/src/Cardano/Wallet/Deposit/Pure.hs

```diff
@@ -0,0 +1,192 @@
+
+-- depend on the private key only, not on the entire wallet state
+deriveAddress :: WalletState -> (Customer -> Address)
+deriveAddress = undefined
+
+knownCustomer :: Customer -> WalletState -> Bool
+knownCustomer c = (c `Map.member`) . customers
+
+knownCustomerAddress :: Address -> WalletState -> Bool
+knownCustomerAddress address = isJust . isCustomerAddress address
+
+isCustomerAddress :: Address -> WalletState -> Maybe Customer
+isCustomerAddress _ _ = Nothing
+
+{-----------------------------------------------------------------------------
+    Operations
+    Reading from the blockchain
+------------------------------------------------------------------------------}
+
+fromGenesis :: Read.GenesisData -> WalletState
+fromGenesis = undefined
+
+fromGenesisUTxO :: Read.UTxO -> WalletState
+fromGenesisUTxO utxo =
+    WalletState
+        { customers = Map.empty
+        , changeAddress = Read.dummyAddress
+        , utxoHistory = UTxOHistory.empty utxo
+        , submissions = Sbm.empty
+        }
+
+localTip :: WalletState -> Read.ChainPoint
+localTip = error "localTip"
+
+rollForwardMany :: NonEmpty Read.Block -> WalletState -> WalletState
+rollForwardMany blocks w = foldl' (flip rollForwardOne) w blocks
+
+rollForwardOne :: Read.Block -> WalletState -> WalletState
+rollForwardOne block w =
+    w
+        { utxoHistory = rollForwardUTxO isOurs block (utxoHistory w)
+        , submissions = Delta.apply (Sbm.rollForward block) (submissions w)
+        }
+  where
+    isOurs :: Address -> Bool
+    isOurs addr =
+        ( addr == changeAddress w ) || knownCustomerAddress addr w
+        -- FIXME: Consider payment part only, ignore staking part.
+
+rollForwardUTxO
+    :: (Address -> Bool) -> Read.Block -> UTxOHistory -> UTxOHistory
+rollForwardUTxO isOurs block u =
+    Delta.apply (UTxOHistory.AppendBlock slot deltaUTxO) u
+  where
+    (deltaUTxO,_) = Balance.applyBlock isOurs block (UTxOHistory.getUTxO u)
+    slot = Read.slot . Read.blockHeaderBody $ Read.blockHeader block
+
+rollBackward
+    :: Read.ChainPoint
+    -> WalletState
+    -> (WalletState, Read.ChainPoint)
+rollBackward = undefined
+
+availableBalance :: WalletState -> Read.Value
+availableBalance w =
+    UTxO.balance $ Balance.availableUTxO utxo pending
+  where
+    pending = listTxsInSubmission w
+    utxo = UTxOHistory.getUTxO $ utxoHistory w
+
+data TxSummary = TxSummary
+    { txid :: Read.TxId
+    , blockHeaderBody :: Read.BHBody
+    , transfer :: ValueTransfer
+    }
+
+data ValueTransfer = ValueTransfer
+    { spent :: Read.Value
+    , received :: Read.Value
+    }
+
+getCustomerHistory :: Customer -> WalletState -> [TxSummary]
+getCustomerHistory = undefined
+
+{-----------------------------------------------------------------------------
+    Operations
+    Writing to blockchain
+------------------------------------------------------------------------------}
+
+createPayment :: [(Address, Write.Value)] -> WalletState -> Maybe Write.Tx
```

Comment:
Note that we also need the following from the "local node tip state":
- `PParams`
- `EpochInfo` / `TimeTranslation`

Reply:
Right. I would put this information into the `WalletState` type, and update it during `rollForward*`. This design is more atomic than our previous versions, and also allows us to keep the type signature `createPayment` as written.

Reply:
My concern would be that retrieving `PParam, EpochInfo` _exactly_ in sync with the `WalletState` would be too complicated, and also not really provide any benefit.

Reply:
👍 I looked up the treatment of `PParam` updates in the ledger specification, it's indeed the case that the acceptance of an update proposal is not recorded in any transaction, but "happens" in the ledger state. Acceptance does involve the treasury, and it's not easy to tell what the `PParams` are without knowing a lot about the current ledger state.

Still, using functions such as

```agda
setPParams : PParams → WalletState → WalletState
setEpochInfo : EpochInfo → WalletState → WalletState
```

it's possible to make `PParams` and `EpochInfo` part of the ledger state; the question is whether this is worth doing as opposed to adding them as parameters to `createPayment`.